File: animations-java/src/main/java/dev/hogoshi/animations/bezier/Point.java
package dev.hogoshi.animations.bezier;

/**
 * Data class which allows to easily manipulate with 2D points
 *
 * @author hogoshi
 */
public class Point {

    private double x;
    private double y;

    public Point() {
        setX(0);
        setY(0);
    }

    public Point(double x, double y) {
        setX(x);
        setY(y);
    }

    public Point(Point point) {
        setX(point.getX());
        setY(point.getY());
    }

    public Point copy() {
        return new Point(this);
    }

    public Point multiply(double x, double y) {
        setX(getX() * x);
        setY(getY() * y);

        return this;
    }

    public Point multiply(double factor) {
        return multiply(factor, factor);
    }

    public Point multiply(Point point) {
        return multiply(point.getX(), point.getY());
    }

    public Point add(double x, double y) {
        setX(getX() + x);
        setY(getY() + y);

        return this;
    }

    public Point add(Point point) {
        return add(point.getX(), point.getY());
    }

    public Point add(double addition) {
        return add(addition, addition);
    }

    public Point subtract(double x, double y) {
        setX(getX() - x);
        setY(getY() - y);

        return this;
    }

    public Point subtract(Point point) {
        return add(point.getX(), point.getY());
    }

    public Point subtract(double subtraction) {
        return add(subtraction, subtraction);
    }

    public Point divide(double x, double y) {
        setX(getX() / x);
        setY(getY() / y);

        return this;
    }

    public Point divide(Point point) {
        return divide(point.getX(), point.getY());
    }

    public Point divide(double division) {
        return divide(division, division);
    }

    public Point set(double x, double y) {
        setX(x);
        setY(y);

        return this;
    }

    public Point set(Point point) {
        return set(point.getX(), point.getY());
    }

    public double getX() {
        return x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return y;
    }

    public void setY(double y) {
        this.y = y;
    }

}

File: animations-kotlin/src/main/kotlin/dev/hogoshi/animations/bezier/Point.kt
package dev.hogoshi.animations.bezier

/**
 * Data class which allows to easily manipulate with 2D points
 * @author hogoshi
 */
data class Point(
    var x: Double = 0.0,
    var y: Double = 0.0
) {

    constructor(point: Point) : this(point.x, point.y)

    constructor(str: String) : this() {
        val str = str.replace(" ", "")
        if (!str.contains(",")) return

        val splitted = str.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (splitted.size <= 1) return

        try {
            this.x = splitted[0].trim { it <= ' ' }.replace("\uFEFF", "").toDouble()
            this.y = splitted[1].trim { it <= ' ' }.replace("\uFEFF", "").toDouble()
        } catch (e: Exception) {
            throw e
        }
    }

    fun copy(): Point {
        return Point(this)
    }

    fun multiply(x: Double, y: Double): Point {
        this.x *= x
        this.y *= y

        return this
    }

    fun multiply(factor: Double): Point {
        return multiply(factor, factor)
    }

    fun multiply(point: Point): Point {
        return multiply(point.x, point.y)
    }

    fun add(x: Double, y: Double): Point {
        this.x += x
        this.y += y

        return this
    }

    fun add(point: Point): Point {
        return add(point.x, point.y)
    }

    fun add(addition: Double): Point {
        return add(addition, addition)
    }

    fun subtract(x: Double, y: Double): Point {
        this.x -= x
        this.y -= y

        return this
    }

    fun subtract(point: Point): Point {
        return add(point.x, point.y)
    }

    fun subtract(subtraction: Double): Point {
        return add(subtraction, subtraction)
    }

    fun divide(x: Double, y: Double): Point {
        this.x /= x
        this.y /= y

        return this
    }

    fun divide(point: Point): Point {
        return divide(point.x, point.y)
    }

    fun divide(division: Double): Point {
        return divide(division, division)
    }

    fun set(x: Double, y: Double): Point {
        this.x = x
        this.y = y

        return this
    }

    fun set(point: Point): Point {
        return set(point.x, point.y)
    }

}